File: src/main/java/com/vaadin/starter/beveragebuddy/ui/MainLayout.java
/*
 * Copyright 2000-2017 Vaadin Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package com.vaadin.starter.beveragebuddy.ui;

import com.vaadin.flow.component.Text;
import com.vaadin.flow.component.dependency.HtmlImport;
import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.component.html.H2;
import com.vaadin.flow.component.icon.Icon;
import com.vaadin.flow.component.icon.VaadinIcons;
import com.vaadin.flow.component.page.Viewport;
import com.vaadin.flow.router.AfterNavigationEvent;
import com.vaadin.flow.router.AfterNavigationObserver;
import com.vaadin.flow.router.RouterLayout;
import com.vaadin.flow.router.RouterLink;
import com.vaadin.flow.server.InitialPageSettings;
import com.vaadin.flow.server.PageConfigurator;
import com.vaadin.starter.beveragebuddy.ui.views.categorieslist.CategoriesList;
import com.vaadin.starter.beveragebuddy.ui.views.reviewslist.ReviewsList;

/**
 * The main layout contains the header with the navigation buttons, and the
 * child views below that.
 */
@HtmlImport("frontend://styles/shared-styles.html")
@Viewport("width=device-width, minimum-scale=1.0, initial-scale=1.0, user-scalable=yes")
public class MainLayout extends Div
        implements RouterLayout, AfterNavigationObserver, PageConfigurator {

    private static final String ACTIVE_ITEM_STYLE = "main-layout__nav-item--selected";
    private RouterLink categories;
    private RouterLink reviews;

    public MainLayout() {
        H2 title = new H2("Beverage Buddy");
        title.addClassName("main-layout__title");

        reviews = new RouterLink(null, ReviewsList.class);
        reviews.add(new Icon(VaadinIcons.LIST), new Text("Reviews"));
        reviews.addClassName("main-layout__nav-item");

        categories = new RouterLink(null, CategoriesList.class);
        categories.add(new Icon(VaadinIcons.ARCHIVES), new Text("Categories"));
        categories.addClassName("main-layout__nav-item");

        Div navigation = new Div(reviews, categories);
        navigation.addClassName("main-layout__nav");

        Div header = new Div(title, navigation);
        header.addClassName("main-layout__header");
        add(header);

        addClassName("main-layout");
    }

    @Override
    public void afterNavigation(AfterNavigationEvent event) {
        // updating the active menu item based on if either of views is active
        // (note that this is triggered even for the error view)
        String segment = event.getLocation().getFirstSegment();
        boolean reviewsActive = segment.equals(reviews.getHref());
        boolean categoriesActive = segment.equals(categories.getHref());

        reviews.setClassName(ACTIVE_ITEM_STYLE, reviewsActive);
        categories.setClassName(ACTIVE_ITEM_STYLE, categoriesActive);
    }

    @Override
    public void configurePage(InitialPageSettings settings) {
        settings.addMetaTag("apple-mobile-web-app-capable", "yes");
        settings.addMetaTag("apple-mobile-web-app-status-bar-style", "black");
    }
}

File: frontend/src/main-layout.ts
import {LitElement, html, customElement, css, unsafeCSS, CSSResult} from 'lit-element';
import '@polymer/iron-icon';
import '@vaadin/vaadin-icons';

// FIXME: import .css files without @ts-ignore (https://github.com/vaadin/flow/issues/6335)
//@ts-ignore
import mainLayoutStyles from '../styles/main-layout-styles.css';
const styles : CSSResult = unsafeCSS(mainLayoutStyles);

@customElement('main-layout')
export class MainLayout extends LitElement {

    static styles = css`${styles}`;

    render() {
        return html`
          <div class="main-layout__header">
            <h2 class="main-layout__title">Beverage Buddy</h2>
            <div class="main-layout__nav">
              <a class="main-layout__nav-item" href=""><iron-icon icon="vaadin:list"></iron-icon>Reviews</a>
              <a class="main-layout__nav-item" href="categories"><iron-icon icon="vaadin:archives"></iron-icon>Categories</a>
              <a class="main-layout__nav-item" href="client-reviews"><iron-icon icon="vaadin:list"></iron-icon>Connect-Reviews</a>
              <a class="main-layout__nav-item" href="client-categories"><iron-icon icon="vaadin:archives"></iron-icon>Connect-Categories</a>
            </div>
          </div>
          <slot></slot>
        `;
    }
}
